### PR TITLE
Add support for saving predictions during training and validation

### PR DESCRIPTION
**Pull request by nagyrajmund**
# Changes

1. Fixed a bug where the yaml config files were not at all read because of a clash between configargparse and Trainer.add_argparse_args()
2. Introduce the functionality to save model predictions during training/validation. This is useful because the training/validation losses don't accurately measure the model performance, so the generated gestures can be used for tracking it.
3. The gesture generation on semantic and random test segments can be enabled now with the `--generate_semantic_test_predictions` and `generate_random_test_predictions` config flags.
4. Updated README for obj_evaluation.

### CODE REVIEW

File: gesticulator/train.py [hunk 1]
```diff
@@ -16,48 +17,24 @@
...
+    logger = create_logger(model.save_dir)
 
-    # DEFAULTS used by Trainer
-    early_stop_callback = EarlyStopping(
-        monitor='avg_val_loss',
-        min_delta=0.00,
-        patience=5,
-        verbose=False,
-        mode='min'
-    )
```
Comment (nagyrajmund) [inline gesticulator/train.py:27]: I removed early stopping because we don't have any metric that can be reliably used for this purpose. YET. :)

File: gesticulator/train.py [hunk 2]
```diff
@@ -16,48 +17,24 @@
...
+    trainer = Trainer.from_argparse_args(hparams, logger=logger)
 
     if not hparams.no_train:
         trainer.fit(model)
     
-    # Save the model
-    save_path  = os.path.join(model.save_dir, 'trained_model_data')
-    print(f"Saving the model to {save_path}...")
-    model_data = {'state_dict' : model.state_dict(), 'hparams': model.hyper_params}
-    torch.save(model_data, save_path)
```
Comment (nagyrajmund) [inline gesticulator/train.py:38]: Model saving will be added later.

File: gesticulator/train.py [hunk 3]
```diff
@@ -16,48 +17,24 @@
...
-def save_videos(save_dir, run_name):
-    """Generate the gesticulation videos for a test sequence."""
-    raw_gesture_path = os.path.join(save_dir, 'test_videos/raw_data')
-    output_dir = os.path.join(save_dir, 'test_videos')
-    data_pipe = 'utils/data_pipe.sav'
-
-    generate_videos(raw_input_folder=raw_gesture_path,
-                    output_folder=output_dir, 
-                    run_name=run_name,
-                    data_pipe_dir=data_pipe)
```
Comment (nagyrajmund) [inline gesticulator/train.py:59]: This functionality is now in the model.

File: gesticulator/config/default_model_config.yaml [hunk 4]
```diff
@@ -1,40 +1,48 @@
...
+# The training loss is MSE(motion_pred, motion_orig) + vel_coef * MSE(velocity_pred, velocity_orig)
+vel_coef: 0.6 
+dropout: 0.2
+dropout_multiplier: 4.0 # The dropout is multiplied by this factor in the conditioning layer
+
+# ---- Parameters for saving model predictions ----
+save_val_predictions_every_n_epoch: 0
+save_train_predictions_every_n_epoch: 0
+saved_prediction_duration_sec: 9
+prediction_save_formats: ["video"] # Can be "bvh_file" and "raw_gesture" too
```
Comment (nagyrajmund) [inline gesticulator/config/default_model_config.yaml:39]: These are the new config parameters for saving the gestures

File: gesticulator/model/model.py [hunk 5]
```diff
@@ -430,12 +442,19 @@ def training_step(self, batch, batch_nb):
...
 
         return output
 
-    
+    def training_epoch_end(self, outputs):
+        elapsed_epochs = self.current_epoch - self.last_saved_train_prediction_epoch 
+        
+        if elapsed_epochs >= self.hyper_params.save_train_predictions_every_n_epoch:
+            self.last_saved_train_prediction_epoch = self.current_epoch
+            self.generate_training_predictions()
```
Comment (nagyrajmund) [inline gesticulator/model/model.py:454]: These generate...() functions are defined in the PredictionSavingMixin in prediction_saving.py!

File: gesticulator/model/prediction_saving.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+      
+        self.save_prediction(predicted_gestures, "validation")
+
+    def generate_semantic_test_predictions(self):
+        """Generate gestures for the 7 chosen semantic test inputs, and save the results."""
+        print("\nGeneratic semantic test gestures:", flush=True)
+        
+        audio_full, text_full = self.semantic_test_input
+        
+        semantic_start_times = [55, 150, 215, 258, 320, 520, 531]
```
Comment (nagyrajmund) [inline gesticulator/model/prediction_saving.py:103]: The semantic test segments from the other file (NaturalTalking_05) can be added later (are they necessary?)

File: gesticulator/train.py (hunk 1 above)
Comment (Svito-zar) [reply gesticulator/train.py:27]: You are right, but that's what we used in the experiments for the paper. So I think it should stay here

File: gesticulator/config/default_model_config.yaml (hunk 4 above)
Comment (Svito-zar) [reply gesticulator/config/default_model_config.yaml:39]: great!

File: gesticulator/model/prediction_saving.py (hunk 6 above)
Comment (Svito-zar) [reply gesticulator/model/prediction_saving.py:103]: well, they were a part of the final evaluation and the user study, so they should be added eventually

File: gesticulator/train.py (hunk 1 above)
Comment (nagyrajmund) [reply gesticulator/train.py:27]: Early stopping was [already disabled when we released the repository](https://github.com/Svito-zar/gesticulator/blob/3c8112bff9916519c386e67b20d2f1319bb700a9/gesticulator/train.py#L30), and if my memory serves right you also mentioned that the experiments ran for a fixed number of epochs. (The paper doesn't mention early stopping either.)

So actually this piece of code was never used anywhere, it's just a leftover that should have been long removed :smiley:

File: gesticulator/model/prediction_saving.py (hunk 6 above)
Comment (nagyrajmund) [reply gesticulator/model/prediction_saving.py:103]: In that case I will add them to this PR :)
